apiextensions: add integration test for name conflicts

nikhita opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: Add integration test for name conflicts.
Create 2 CRDs with name conflict. The controller sets the NameConflict condition on the second one. Delete the first one. The second CRD should now get accepted.

Update:
- [x] Add integration test for name conflicts
- [x] Fix naming controller so that when a resource is deleted, other resources in the same group are added to the queue
- [x]  Add integration test for self link (cluster scoped resources)
- [x] Add bigger poll interval
- [x] Fix DeleteCustomResourceDefinition poll condition


**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: for #45511 

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
NONE
```
@sttts

nikhita commented on staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go:
```diff
@@ -379,3 +379,38 @@ func TestPreserveInt(t *testing.T) {
+
+	// TODO: Make it pass here
+	if crd.Status.Conditions[0].Status == "True" {
+		t.Fatalf("error: names chosen for this CustomResourceDefinition conflict with others in the group")
```
When I create 2 CRDs with name conflicts, delete one of them,  _wait_ and then check the other, the name conflict condition is removed.
However in the tests, no waiting occurs and this fails. What can I do to fix this?

sttts replied:
See above: `wait.Poll`

nikhita commented on staging/src/k8s.io/kube-apiextensions-server/test/integration/testserver/resources.go:
```diff
@@ -100,6 +100,24 @@ func NewCurletInstance(namespace, name string) *unstructured.Unstructured {
 	}
 }
 
+func NewNoxu2CustomResourceDefinition(scope apiextensionsv1alpha1.ResourceScope) *apiextensionsv1alpha1.CustomResourceDefinition {
```
Better naming suggestions welcome.

sttts commented:
@k8s-bot ok to test

sttts commented on staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go:
```diff
@@ -379,3 +379,38 @@ func TestPreserveInt(t *testing.T) {
+	}
+
+	// TODO: Make it pass here
+	if crd.Status.Conditions[0].Status == "True" {
```
Don't we need polling around this to avoid that the test races with the controller?

sttts replied:
`Conditions[0]` is not safe. There might be more conditions. We have to go through all conditions to find the right one. @deads2k wrote some helpers (for the internal version only though I fear).

sttts commented on staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go:
```diff
@@ -379,3 +379,38 @@ func TestPreserveInt(t *testing.T) {
+
+	noxu2Definition := testserver.NewNoxu2CustomResourceDefinition(apiextensionsv1alpha1.NamespaceScoped)
+	_, err = testserver.CreateNewCustomResourceDefinition(noxu2Definition, apiExtensionClient, clientPool)
+	if err == nil {
```
This takes 30sec. I would prefer to ignore discovery here and and just do a `wait.Poll` checking the status for the `NameConflict`.

sttts commented on staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go:
```diff
@@ -379,3 +380,45 @@ func TestPreserveInt(t *testing.T) {
+		}
+
+		for _, condition := range crd.Status.Conditions {
+			if condition.Status == "False" {
```
`if condition.Type == CustomResourceDefinitionConditionType && condition.Status == "False"`

sttts replied:
You might hit the terminating condition here.

nikhita commented:
So the problem (as discussed on slack) was that when a resource was deleted, other resources in the same group were not added to the queue. This fixes that.

nikhita commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -214,6 +214,30 @@ func (c *NamingConditionController) sync(key string) error {
+	for _, condition := range crd.Status.Conditions {
+		if condition.Type == "Terminating" && condition.Status == "False" {
+			// wait for all accepted names to be updated
+			time.Sleep(1 * time.Second)
```
Not sure if `time.Sleep` is the best way to do this here. This is needed because it takes time for the list of all accepted names to be updated after deletion.

nikhita commented:
@k8s-bot pull-kubernetes-unit test this

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -214,6 +214,30 @@ func (c *NamingConditionController) sync(key string) error {
+
+	// when a resource is deleted, enqueue other resources of the same group
+	for _, condition := range crd.Status.Conditions {
+		if condition.Type == "Terminating" && condition.Status == "False" {
```
there is a constant for `"Terminating"` and one for "False"

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -214,6 +214,30 @@ func (c *NamingConditionController) sync(key string) error {
+			}
+			for _, curr := range list {
+				if curr.Spec.Group == crd.Spec.Group {
+					if curr.Name != crd.Name {
```
can be one if clause

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -214,6 +214,30 @@ func (c *NamingConditionController) sync(key string) error {
+			if err != nil {
+				return nil
+			}
+			for _, curr := range list {
```
can we turn this for loop into a `RequeueAllOtherGroupCRDs`? We can then reuse it at the bottom of the func.

nikhita committed:
Add integration test for apiextensions

* add integration test for name conflicts of CRDs
* fix naming controller so that when a resource is deleted,
other resources in the same group are added to the queue
* add integration test for self link (clusterscoped resources)
* fix minor nits

address comments
add comments and fix nits

Bigger poll interval

customResource -> crd

Fix DeleteCustomResourceDefinition condition

nikhita commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

nikhita commented:
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

nikhita commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -358,3 +373,16 @@ func (c *NamingConditionController) deleteCustomResourceDefinition(obj interface
 	c.enqueue(castObj)
 }
+
+func (c *NamingConditionController) requeueAllOtherGroupCRDs(crd *apiextensions.CustomResourceDefinition) error {
```
:+1:

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -264,14 +284,9 @@ func (c *NamingConditionController) sync(key string) error {
-	list, err := c.crdLister.List(labels.Everything())
+	err = c.requeueAllOtherGroupCRDs(crd)
 	if err != nil {
-		return nil
```
wondering why we had a `return nil` here.

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -264,14 +284,9 @@ func (c *NamingConditionController) sync(key string) error {
 	// we updated our status, so we may be releasing a name.  When this happens, we need to rekick everything in our group
 	// if we fail to rekick, just return as normal.  We'll get everything on a resync
-	list, err := c.crdLister.List(labels.Everything())
+	err = c.requeueAllOtherGroupCRDs(crd)
```
nit: `if err := ...; err != nil {`

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -236,6 +236,31 @@ func (c *NamingConditionController) sync(key string) error {
+		return err
+	}
+
+	for _, condition := range crd.Status.Conditions {
```
Not sure I understand the logic behind this loop. I would have expected a `c.requeueAllOtherGroupCRDs(crd)` in line 233 above. Wouldn't that be enough? Of course, we would need some other signature for `requeueAllOtherGroupCRDs` which works with the key (what's the key? namespace/name?)

nikhita replied:
Oh yes, that should be enough. Key is the name. I can extract the group name from the key and requeue the CRDs in that group.

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -358,3 +356,17 @@ func (c *NamingConditionController) deleteCustomResourceDefinition(obj interface
 }
+
+func (c *NamingConditionController) requeueAllOtherGroupCRDs(name string) error {
+	group := strings.SplitAfterN(name, ".", 2)
```
better use `pluralGroup :=`. And I think `SplitN` should be fine as well.

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -230,12 +231,20 @@ func equalToAcceptedOrFresh(requestedName, acceptedName string, usedNames sets.S
 		return err
 	}
 
+	crd := &apiextensions.CustomResourceDefinition{}
```
can we move the deepcopy back to the old place?

sttts commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -230,12 +231,20 @@ func equalToAcceptedOrFresh(requestedName, acceptedName string, usedNames sets.S
 func (c *NamingConditionController) sync(key string) error {
 	inCustomResourceDefinition, err := c.crdLister.Get(key)
 	if apierrors.IsNotFound(err) {
+		if err := c.requeueAllOtherGroupCRDs(key); err != nil {
```
deserves a comment: CRD was deleted and has freed its names. Reconsider all other CRDs in the same group.

sttts commented:
/lgtm

nikhita commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

nikhita commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

sttts commented:
/lgtm

sttts commented:
@k8s-bot pull-kubernetes-node-e2e test this

nikhita commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this
@k8s-bot pull-kubernetes-node-e2e test this

nikhita commented:
@k8s-bot pull-kubernetes-node-e2e test this
@k8s-bot pull-kubernetes-e2e-kops-aws test this

deads2k commented on staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go:
```diff
@@ -230,6 +231,11 @@ func equalToAcceptedOrFresh(requestedName, acceptedName string, usedNames sets.S
 	if apierrors.IsNotFound(err) {
+		// CRD was deleted and has freed its names.
+		// Reconsider all other CRDs in the same group.
+		if err := c.requeueAllOtherGroupCRDs(key); err != nil {
```
In my head, I would have expected this on the deletion handling function.

deads2k replied:
Oh, I see.  That makes sense.

deads2k commented on staging/src/k8s.io/kube-apiextensions-server/test/integration/testserver/resources.go:
```diff
@@ -100,20 +118,20 @@ func NewCurletInstance(namespace, name string) *unstructured.Unstructured {
 	// wait until the resource appears in discovery
-	err = wait.PollImmediate(30*time.Millisecond, 30*time.Second, func() (bool, error) {
-		resourceList, err := apiExtensionsClient.Discovery().ServerResourcesForGroupVersion(customResourceDefinition.Spec.Group + "/" + customResourceDefinition.Spec.Version)
+	err = wait.PollImmediate(500*time.Millisecond, 30*time.Second, func() (bool, error) {
```
why slow down the polling cycle?  Are we running this slow?

sttts replied:
30sec gives you a lot of audit spam with 30msec interval. Wondering, was audit always enabled in integration tests?
